core/src/store/alter_table.rs

```diff
@@ -28,41 +27,168 @@ pub enum AlterTableError {
 
     #[error("Schemaless table does not support ALTER TABLE: {0}")]
     SchemalessTableFound(String),
+
+    #[error("conflict - Vec expected but Map row found")]
+    ConflictOnUnexpectedMapRowFound,
 }
 
 #[async_trait(?Send)]
-pub trait AlterTable {
-    async fn rename_schema(&mut self, _table_name: &str, _new_table_name: &str) -> Result<()> {
-        let msg = "[Storage] AlterTable::rename_schema is not supported".to_owned();
+pub trait AlterTable: Store + StoreMut {
+    async fn rename_schema(&mut self, table_name: &str, new_table_name: &str) -> Result<()> {
+        let mut schema = self
+            .fetch_schema(table_name)
+            .await?
+            .ok_or_else(|| AlterTableError::TableNotFound(table_name.to_owned()))?;
+        schema.table_name = new_table_name.to_owned();
+        self.insert_schema(&schema).await?;
+
+        let rows = self
+            .scan_data(table_name)
+            .await?
+            .try_collect::<Vec<_>>()
+            .await?;
 
-        Err(Error::StorageMsg(msg))
+        self.insert_data(new_table_name, rows).await?;
+        self.delete_schema(table_name).await
     }
 
     async fn rename_column(
         &mut self,
-        _table_name: &str,
-        _old_column_name: &str,
-        _new_column_name: &str,
+        table_name: &str,
+        old_column_name: &str,
+        new_column_name: &str,
     ) -> Result<()> {
-        let msg = "[Storage] AlterTable::rename_column is not supported".to_owned();
+        let mut schema = self
+            .fetch_schema(table_name)
+            .await?
+            .ok_or_else(|| AlterTableError::TableNotFound(table_name.to_owned()))?;
+
+        let column_defs = schema
+            .column_defs
+            .as_mut()
+            .ok_or_else(|| AlterTableError::SchemalessTableFound(table_name.to_owned()))?;
+
+        if column_defs
+            .iter()
+            .any(|column_def| column_def.name == new_column_name)
+        {
+            return Err(AlterTableError::AlreadyExistingColumn(new_column_name.to_owned()).into());
+        }
+
+        column_defs
+            .iter_mut()
+            .find(|column_def| column_def.name == old_column_name)
+            .ok_or_else(|| AlterTableError::RenamingColumnNotFound)?
+            .name = new_column_name.to_owned();
 
-        Err(Error::StorageMsg(msg))
+        let rows = self
+            .scan_data(table_name)
+            .await?
+            .try_collect::<Vec<_>>()
+            .await?;
+
+        self.delete_schema(table_name).await?;
```

Comment:
I tried to remove this line cuz both `insert_schema` and `insert_data` already have `overwirting` meaning but the only failure is `git + file storage` 😢

Reply:
```
[RUN] ALTER TABLE Bar RENAME COLUMN id TO new_id
>>>: cd "tmp/git_storage_file/alter_table_rename" && "git" "add" "."
+++: Output { status: ExitStatus(unix_wait_status(0)), stdout: "", stderr: "" }
>>>: cd "tmp/git_storage_file/alter_table_rename" && "git" "commit" "-m" "[GitStorage::insert_schema] Bar"
+++: Output { status: ExitStatus(unix_wait_status(0)), stdout: "[main e444acc] [GitStorage::insert_schema] Bar\n 1 file changed, 1 insertion(+), 1 deletion(-)\n", stderr: "" }
>>>: cd "tmp/git_storage_file/alter_table_rename" && "git" "add" "."
+++: Output { status: ExitStatus(unix_wait_status(0)), stdout: "", stderr: "" }
>>>: cd "tmp/git_storage_file/alter_table_rename" && "git" "commit" "-m" "[GitStorage::insert_data] Bar - 3 rows"
+++: Output { status: ExitStatus(unix_wait_status(256)), stdout: "On branch main\nnothing to commit, working tree clean\n", stderr: "" }
```

```
[RUN] ALTER TABLE Bar RENAME COLUMN id TO new_id
>>>: cd "tmp/git_storage_json/alter_table_rename" && "git" "add" "."
+++: Output { status: ExitStatus(unix_wait_status(0)), stdout: "", stderr: "" }
>>>: cd "tmp/git_storage_json/alter_table_rename" && "git" "commit" "-m" "[GitStorage::delete_schema] {table_name}"
+++: Output { status: ExitStatus(unix_wait_status(0)), stdout: "[main ea9e3f7] [GitStorage::delete_schema] {table_name}\n 2 files changed, 4 deletions(-)\n delete mode 100644 Bar.jsonl\n delete mode 100644 Bar.sql\n", stderr: "" }
>>>: cd "tmp/git_storage_json/alter_table_rename" && "git" "add" "."
+++: Output { status: ExitStatus(unix_wait_status(0)), stdout: "", stderr: "" }
>>>: cd "tmp/git_storage_json/alter_table_rename" && "git" "commit" "-m" "[GitStorage::insert_schema] Bar"
+++: Output { status: ExitStatus(unix_wait_status(0)), stdout: "[main d78f0a0] [GitStorage::insert_schema] Bar\n 2 files changed, 1 insertion(+)\n create mode 100644 Bar.jsonl\n create mode 100644 Bar.sql\n", stderr: "" }
>>>: cd "tmp/git_storage_json/alter_table_rename" && "git" "add" "."
+++: Output { status: ExitStatus(unix_wait_status(0)), stdout: "", stderr: "" }
>>>: cd "tmp/git_storage_json/alter_table_rename" && "git" "commit" "-m" "[GitStorage::insert_data] Bar - 3 rows"
+++: Output { status: ExitStatus(unix_wait_status(0)), stdout: "[main b1b0a62] [GitStorage::insert_data] Bar - 3 rows\n 1 file changed, 3 insertions(+)\n", stderr: "" }
```

Reply:
that really makes sense. it would be enough to remove `self.delete_schema`